rfcs/records.md

```diff
@@ -0,0 +1,196 @@
+introduce support for packed records where the table definition must use types and writes that don't abide by these types will trigger an error.
+
+> TODO: It's going to be difficult to migrate to packed records. Is there anything we can do right now to keep this possibility open without enforcing types at runtime?
+
+The rest of this proposal goes into syntax and semantics. The goal of this proposal is to solve the problem of object storage both for simple objects and for classes -- that is, if we add records we won't need to add classes.
+
+### Record type
+
+A record is a collectable object that stores the field values as a inline array of values (TValues) as well as a pointer to the shape. Shape is a table
+that stores various lookup data as implementation details as well as metafields. `type(r)` is `"record"`; `getmetatable(r)` can be used to retrieve the shape.
+
+Being a collectable object, records use raw equality by default when comparing using `==` or hashing; equality behavior can be overridden via `__eq`. From this
+perspective, one could think of records as a user-defined userdata type as opposed to a host-defined userdata type: both typically expose a strict set of fields,
+both are heap-allocated, both use contiguous storage.
+
+Shape contains field lookup data in extra storage that's only allocated for shape tables, as well as the regular table fields. The field lookup data is internal and
+immutable; for example, it might contain a string->index dictionary to be able to quickly locate fields in internal storage.
+
+> TODO: Still not fully set on whether we can get by without a first class shape type.
+
+Reading and writing fields from a record uses `.` or `[]` operator; like tables, passing the field name returns the field value. Unlike tables, if the field is not
+present in the table, the error is raised. This is in constrast with tables where `nil` is returned for unknown keys upon read; records are meant to be stricted than
+tables and as such returning `nil` will mask valuable errors, and make it more difficult to be strict about the types of the result.
+
+Invoking methods with `:` desugars into `getmetatable(obj).__index.method(obj, args)` instead of the usual `obj.method(obj, args)`. This is important because it allows to
+keep the method calls as efficient as possible, as they don't need to check whether the object has a given method.
+
+> TODO: Should we use `__namecall` instead of `__index`? It seems more consistent, but at the same time `__namecall` today expects a function so it might be best
+> to leave it as is?
+
+> TODO: How do we expose the record keys? Should it be a builtin? Accessible through shape? Not available initially?
+
+### Defining and constructing records
+
+To define a record, you need to create the shape, which you can do using the newly introduced syntax with a context-sensitive `record` keyword:
+
+Syntax A:
+
+```
+record Person = { name: string, age: number }
```

Comment:
How about something like
```suggestion
record Person
    name: string
    age: number
end
```
This looks a little bit truer to stock Lua and gives us a place to start talking about nice syntax for methods.

Reply:
That actually looks a lot better to me than the other syntax does, plus it's a lot more unambiguous when it comes to syntax.